crates/sui-adapter/src/adapter.rs

```diff
@@ -992,136 +1009,63 @@ pub fn resolve_and_type_check(
+                    InputObjectKind::ImmOrOwnedMoveObject(ref_),
+                    idx,
+                    param_type,
+                    objects,
+                    &mut object_data,
+                    &mut mutable_ref_objects,
+                    &mut by_value_objects,
+                    &mut object_type_map,
+                )?,
+                CallArg::Object(ObjectArg::SharedObject(id)) => serialize_object(
+                    InputObjectKind::SharedMoveObject(id),
+                    idx,
+                    param_type,
+                    objects,
+                    &mut object_data,
+                    &mut mutable_ref_objects,
+                    &mut by_value_objects,
+                    &mut object_type_map,
+                )?,
+                CallArg::ObjVec(vec) => {
+                    if vec.is_empty() {
+                        // bcs representation of the empty vector
+                        return Ok(vec![0]);
                     }
-                    by_value_objects.insert(id);
-                    t
-                }
-                t => {
-                    return Err(ExecutionError::new_with_source(
-                        ExecutionErrorKind::entry_argument_error(
+                    // write length of the vector as uleb128 as it is encoded in BCS and then append
+                    // all (already serialized) object content data
+                    let mut res = vec![];
+                    leb128::write::unsigned(&mut res, vec.len() as u64).unwrap();
+                    let mut element_type = None;
+                    let mut inner_vec_type = None;
+                    for arg in vec {
+                        let object_kind = match arg {
+                            ObjectArg::ImmOrOwnedObject(ref_) => {
+                                InputObjectKind::ImmOrOwnedMoveObject(ref_)
+                            }
+                            ObjectArg::SharedObject(id) => InputObjectKind::SharedMoveObject(id),
+                        };
+                        let (o, arg_type, param_type) = serialize_object(
+                            object_kind,
                             idx,
-                            EntryArgumentErrorKind::TypeMismatch,
-                        ),
-                        format!(
-                            "Found object argument {}, but function expects {:?}",
-                            move_object.type_, t
-                        ),
-                    ));
+                            param_type,
+                            objects,
+                            &mut object_data,
+                            &mut mutable_ref_objects,
+                            &mut by_value_objects,
+                            &mut object_type_map,
+                        )?;
+                        res.extend(o);
+                        element_type = Some(arg_type);
+                        inner_vec_type = Some(param_type);
```

Comment:
maybe a debug assert that if there was a previous value there, it should be the same?

Reply:
This reminded me that I wanted to add corresponding tests but didn't... Also, this should be handled by more than an assertion I think. In fact, we should be checking if each element's type is consistent with vector's signature and report and error (rather than panic) if it's not.